### PR TITLE
Refactor distributed tests

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Description:
Added a new fixture called `distributed` which requesting it by a test, makes the test run for 7 preset distributed configurations.
For now this fixture is used in `test_precision.py` and if was successful, we could use it in other test modules.

### CODE REVIEW

File: tests/ignite/conftest.py [hunk 1]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+    params=[
+        pytest.param(
+            "nccl_gpu",
+            marks=[
+                pytest.mark.distributed,
+                skip_if_has_not_native_dist_support,
+                pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU"),
+            ],
+        ),
+        pytest.param("gloo_cpu_or_gpu", marks=[pytest.mark.distributed, skip_if_has_not_native_dist_support]),
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:366]: Let's split `gloo_cpu_or_gpu` into `gloo_cpu` and `gloo` ?

File: tests/ignite/conftest.py [hunk 2]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+is_horovod_stash_key = pytest.StashKey[bool]()
+is_xla_stash_key = pytest.StashKey[bool]()
+is_xla_single_device_stash_key = pytest.StashKey[bool]()
+
+
+@pytest.fixture(
+    scope="module",
+    params=[
+        pytest.param(
+            "nccl_gpu",
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:359]:
```suggestion
            "nccl",
```

File: tests/ignite/conftest.py [hunk 3]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+            marks=[
+                pytest.mark.distributed,
+                skip_if_has_not_native_dist_support,
+                pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU"),
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:363]: We can create a similar `skip_if_no_gpu = pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")` and reuse it later

File: tests/ignite/conftest.py [hunk 4]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+        pytest.param(
+            "multinode_nccl_gpu",
+            marks=[
+                pytest.mark.multinode_distributed,
+                skip_if_has_not_native_dist_support,
+                pytest.mark.skipif(
+                    "GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed"
+                ),
+            ],
+        ),
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:392]: For now I would remove multi-node configs

File: tests/ignite/conftest.py [hunk 5]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+                skip_if_has_not_xla_support,
+                pytest.mark.skipif(
+                    "NUM_TPU_WORKERS" not in os.environ, reason="Skip if no NUM_TPU_WORKERS in env vars"
+                ),
+            ],
+        ),
+    ],
+)
+def distributed(request, local_rank, world_size):
+    if request.param == "nccl_gpu" or request.param == "gloo_cpu_or_gpu":
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:414]: Let's code such condition as 
```suggestion
    if request.param in ("nccl_gpu", "gloo_cpu_or_gpu"):
```

File: tests/ignite/conftest.py [hunk 6]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+            "rank": multi_node_conf["rank"],
+        }
+        if request.param == "multinode_nccl_gpu":
+            os.environ["MASTER_PORT"] = str(int(os.getenv("MASTER_PORT")) + 1)
+            dist_info["backend"] = "nccl"
+        else:
+            dist_info["backend"] = "gloo"
+
+        yield _create_mnodes_dist_context(dist_info, multi_node_conf)
+        _destroy_mnodes_dist_context()
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:476]: Let's remove multi-node config for this PR and try to enable it later

File: tests/ignite/conftest.py [hunk 7]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+
+    elif request.param == "single_device_xla" or request.param == "xla_nprocs":
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:478]:
```suggestion
    elif request.param in ("single_device_xla", "xla_nprocs"):
```

File: tests/ignite/conftest.py [hunk 8]
```diff
@@ -333,3 +336,159 @@ def _gloo_hvd_execute(func, args, np=1, do_init=False):
...
+)
+skip_if_has_not_xla_support = pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+skip_if_has_not_horovod_support = pytest.mark.skipif(
+    not idist.has_hvd_support, reason="Skip if no Horovod dist support"
+)
+
+
+is_horovod_stash_key = pytest.StashKey[bool]()
+is_xla_stash_key = pytest.StashKey[bool]()
+is_xla_single_device_stash_key = pytest.StashKey[bool]()
```
Comment (vfdev-5) [inline tests/ignite/conftest.py:352]: Please add comments here what are these stash keys, why we are using that and for what.

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot for the PR @sadra-barikbin , overall looks good. I left few comments how to update the code.

File: tests/ignite/conftest.py (hunk 1 above)
Comment (sadra-barikbin) [reply tests/ignite/conftest.py:366]: Do you mean `gloo_cpu` and `gloo_gpu`?

Comment (vfdev-5) [reply tests/ignite/conftest.py]: No, gloo and gloo_cpu. By default nccl and gloo are considered for gpu

**Pull request merged**